### PR TITLE
Update rustc_interface/queries.rs: Cleaner code

### PR DESCRIPTION
**Pull request by apleeagain**
‘Query’ struct: replaced ‘Option’ with ‘OnceCell’ and switched layering of types for much cleaner code in the ‘.compute()’ method. Also more clearly reflects the intent of the struct.

### CODE REVIEW

**Comment by apleeagain** [comment]
How do I add a label? I would like to add the “Code Quality” label

**Comment by apleeagain** [comment]
Regarding ‘QueryResult’:
  - Why is the ‘Steal’ wrapped in a ‘RefCell’?

Furthermore, regarding ‘Steal<T>’ specifically
 - why is there an option to borrow ‘T’ as mutable? That defies the whole purpose of the ‘Steal’.
 - one place where I can find a value of ‘Steal’ to be borrowed mutably is in the above file at around line 65, used to ‘.enter()’ a ‘GlobalCtxt’ wrapped in a ‘Steal’. However, this method does not even require the GlobalCtxt to be borrowed mutably! A ‘.borrow()’ method on the ‘Steal’ would suffice for all purposes above.
 - If the ‘T’ covered by a ‘Steal’ were ever needed to be borrowed mutably, one must simply introduce a ‘.borrow_mut()’ method on ‘Steal’; the Steal itself already conveniently wraps a ‘RefCell’, making this adaptation straightforward.
 
 -> as the ‘Steal’ already functions using its contained ‘RefCell’, wrapping a ‘Steal’ in another ‘RefCell’ as done in the ‘QueryResult’ struct is pointless and a waste of everything. The ‘RefCell’ wrapping could be removed from ‘QueryResult’ as well as the ‘Query’ struct also, simplifying everything. If you can, please help applying these changes as I am currently unable to.

**Comment by apleeagain** [comment]
- Furthermore, avoiding to impart a ‘RefMut’ to the ‘Steal’ in ‘QueryResult’ prevents users from doing stupid things like replacing the Steal with a new Steal. The above suggestions would statically prevent that.

**Comment by fmease** [comment]
> I would like to add the “Code Quality” label

@rustbot label C-cleanup

**Comment by cjgillot** [comment]
> Regarding ‘QueryResult’:
> 
>     * Why is the ‘Steal’ wrapped in a ‘RefCell’?
> 
> 

I agree, we shouldn't need to wrap the Steal in a RefCell.

> Furthermore, regarding ‘Steal’ specifically
> 
>     * why is there an option to borrow ‘T’ as mutable? That defies the whole purpose of the ‘Steal’.
> 

The intent of `Steal` is to save memory by avoiding a clone. It allows some code to borrow the value, until it is finally stolen. It is the thief's responsibility to ensure no-one will need it after it is stolen.

>     * one place where I can find a value of ‘Steal’ to be borrowed mutably is in the above file at around line 65, used to ‘.enter()’ a ‘GlobalCtxt’ wrapped in a ‘Steal’. However, this method does not even require the GlobalCtxt to be borrowed mutably! A ‘.borrow()’ method on the ‘Steal’ would suffice for all purposes above.
> 
>     * If the ‘T’ covered by a ‘Steal’ were ever needed to be borrowed mutably, one must simply introduce a ‘.borrow_mut()’ method on ‘Steal’; the Steal itself already conveniently wraps a ‘RefCell’, making this adaptation straightforward.
> 

That said, having `Steal` here is not necessarily the right data structure. Perhaps this:
```
struct Query<T> {
    inner: RefCell<QueryState<T>>,
}

enum QueryState<T> {
    Empty,
    Value(T),
    Error(SomeErrorType),
    Stolen,
}
```

> 
> -> as the ‘Steal’ already functions using its contained ‘RefCell’, wrapping a ‘Steal’ in another ‘RefCell’ as done in the ‘QueryResult’ struct is pointless and a waste of everything. The ‘RefCell’ wrapping could be removed from ‘QueryResult’ as well as the ‘Query’ struct also, simplifying everything. If you can, please help applying these changes as I am currently unable to.

How can I help?

**Comment by apleeagain** [comment]
I get that the Steal needs to be stolen, and there is a method for that, which all makes sense. But one method on the steal let you borrow the contents, despite a Steal intended to be used as the outcome of a query, and queries should be pure, so you shouldn’t be able to modify its result unless it would be inobservable, which is what the steal method is for.

**Comment by cjgillot** [comment]
I read too fast indeed. I guess the option to borrow mutably should go.

**Comment by apleeagain** [comment]
I like your idea of using an enum. That would remove so many layers of types. Let me have a think.

**Comment by apleeagain** [comment]
The issue is that I find using ‘OnceCell<Result<Steal<T>>>’ more aesthetic. It conveys exactly the purpose of a query: it is computed once, and may result in an error or not. It shows that the value is wrapped in a Steal, which conveys its own information. Each type inherently asserts its own functionality on the resulting whole, each with its own unique and well-understood methods; I find unwrapping and mapping prettier than a match statement.

So on my side, I would stick to my suggestion of converting the ‘Query’ to what I have done, which has already simplified it somewhat.

**Comment by apleeagain** [comment]
@rustbot label S-waiting-on-review

Sorry for my errors, I have fixed them, I think, according to the test outcome diagnostics.

Regarding what you could help me with: you could remove the get_mut() method from Steal, and remove the RefCell from both ‘QueryResult’ as well as ‘Query’, and then work out any necessary refactoring needed in dependent code. I have already marked some areas which would need refactoring, but I haven’t checked in other crates as I do not currently have access to a computer to compile with. Also chec

File: compiler/rustc_interface/src/queries.rs [hunk 1]
```diff
@@ -32,27 +32,36 @@ use std::sync::Arc;
...
+    /// Cell is empty until a value is computed
+    result: OnceCell<Result<RefCell<Steal<T>>>>,
+ //(aplee): OnceCell<Result<Steal<T>>>
 }
+    
+// TODO(aplee): (remove once finished) (see #111668)
+// - remove 'RefCell' both from 'QueryResult' and 'Query',
+// - fix anything which breaks; if needed, add a '.borrow_mut()' method to 'Steal',
+//   although this should be avoided.
+
```
Comment (cjgillot) [inline compiler/rustc_interface/src/queries.rs:44]: Could you finish the PR and remove this?

**Comment by Dylan-DPC** [comment]
@apleeagain we do not support merge commits so kindly revert it, and rebase instead.

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**